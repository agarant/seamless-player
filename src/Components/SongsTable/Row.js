import React from "react";
import { TableCell, TableRow } from "material-ui/Table";
import { HoverHOC } from "../../utils";
import styled from "styled-components";
import PlaylistAdd from "material-ui-icons/PlaylistAdd";
import IconButton from "material-ui/IconButton";
import Menu, { MenuItem } from "material-ui/Menu";

const MyTableRow = styled(TableRow)`
  &:hover {
    cursor: pointer;
  }
`;

const ButtonBox = styled.td`
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  margin-right: 25px;
`;

const ITEM_HEIGHT = 48;

class Row extends React.Component {
  state = {
    anchorEl: null
  };

  addToPlaylistClicked = event => {
    event.stopPropagation();
    this.setState({ anchorEl: event.currentTarget });
  };

  handleRequestClose = () => {
    this.setState({ anchorEl: null });
  };

  rowClicked = () => {
    this.props.playsong(this.props.song);
  };

  addToPlayList = id => e => {
    e.stopPropagation();
    this.props.addToPlaylist(id, this.props.song);
    this.handleRequestClose();
  };

  renderAddButton = () => {
    if (!this.props.hovered) return null;
    const open = Boolean(this.state.anchorEl);

    const { playlists } = this.props;
    return (
      <ButtonBox>
        <IconButton onClick={this.addToPlaylistClicked}>
          <PlaylistAdd />
        </IconButton>
        <Menu
          id="long-menu"
          anchorEl={this.state.anchorEl}
          open={open}
          onRequestClose={this.handleRequestClose}
          PaperProps={{
            style: {
              maxHeight: ITEM_HEIGHT * 4.5,
              width: 200
            }
          }}
        >
          {Object.keys(playlists).map(key => (
            <MenuItem
              key={playlists[key].id}
              onClick={this.addToPlayList(playlists[key].id)}
            >
              {playlists[key].name}
            </MenuItem>
          ))}
        </Menu>
      </ButtonBox>
    );
  };

  render() {
    const { name, artist, album, artwork } = this.props.song;

    return (
      <MyTableRow hover={true} {...this.props} onClick={this.rowClicked}>
        <TableCell style={{ display: "flex", alignItems: "center" }}>
          <img
            src={artwork}
            height={36}
            width={36}
            style={{ marginRight: 10 }}
            alt="artwork"
          />
          {name}
        </TableCell>
        <TableCell>{artist}</TableCell>
        <TableCell>{album}</TableCell>
        <TableCell style={{ position: "relative" }}>
          {this.renderAddButton()}
        </TableCell>
      </MyTableRow>
    );
  }
}

export default HoverHOC(Row);
