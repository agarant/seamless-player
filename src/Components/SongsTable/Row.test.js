import React from "react";
import renderer from "react-test-renderer";
import Row from "./Row";

test("<Row /> renders properly", () => {
  const mockSong = {
    id: "1234",
    name: "Atlas",
    artist: "Luke Christopher",
    album: "TMRW",
    artwork: "mockUrl",
    url: "mockOtherUrl"
  };
  const tree = renderer.create(<Row song={mockSong} playlists={{}} />).toJSON();
  expect(tree).toMatchSnapshot();
});
