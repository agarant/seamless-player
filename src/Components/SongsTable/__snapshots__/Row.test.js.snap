// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> renders properly 1`] = `
<tr
  className="MuiTableRow-root-1 sc-bdVaJa irEbJS"
  hovered={false}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  playlists={Object {}}
  song={
    Object {
      "album": "TMRW",
      "artist": "Luke Christopher",
      "artwork": "mockUrl",
      "id": "1234",
      "name": "Atlas",
      "url": "mockOtherUrl",
    }
  }
>
  <td
    className="MuiTableCell-root-6 MuiTableCell-paddingDefault-9"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <img
      alt="artwork"
      height={36}
      src="mockUrl"
      style={
        Object {
          "marginRight": 10,
        }
      }
      width={36}
    />
    Atlas
  </td>
  <td
    className="MuiTableCell-root-6 MuiTableCell-paddingDefault-9"
  >
    Luke Christopher
  </td>
  <td
    className="MuiTableCell-root-6 MuiTableCell-paddingDefault-9"
  >
    TMRW
  </td>
  <td
    className="MuiTableCell-root-6 MuiTableCell-paddingDefault-9"
    style={
      Object {
        "position": "relative",
      }
    }
  />
</tr>
`;
